### PR TITLE
feat: add MainTopBar + SearchBar components [AR-852] [AR-853] [AR-777]

### PR DESCRIPTION
**Pull request by gongracr**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-852" title="AR-852" target="_blank">SubtaskAR-852</a>  Top AppBar Item
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Description

Added the main navigation components in order to build visually the `MainTopBar`, for that the following was needed:
- Created `SearchBar` (fullscreen and collapsed ones), `UserProfileAvatar` and `UserProfileScreen`
- Created `NavigationUtils` file as a place to keep the shared methods handling the `MainNavigationScreenItems`
- Created the `Dimensions` object to share common reusable dimension units.

### Screenshots 

Captura de pantalla 2022-01-18 a las 13 31 36

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Work flow has automatically linked the PR to a JIRA issue
----

### CODE REVIEW

File: app/src/main/kotlin/com/wire/android/ui/common/Dimensions.kt [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.wire.android.ui.common
+
+import androidx.compose.ui.unit.dp
+
+object Dimensions {
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/common/Dimensions.kt:5]:
```suggestion

@Suppress("MagicNumber")
object Dimensions {
```

Not sure if Detekt will complain about it now, but it definitely will.

File: app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+fun UserProfileAvatar(avatarUrl: String, size: Dp = USER_AVATAR_DEFAULT_SIZE, modifier: Modifier = Modifier, onClick: () -> Unit) {
+    ConstraintLayout(
+        modifier = modifier
+            .padding(10.dp)
+            .size(size)
+    ) {
+        val (avatarImg, statusImg) = createRefs()
+        Image(
+            painter = painterResource(getAvatarAsDrawable(avatarUrl)),
+            contentDescription = "user profile avatar",
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt:33]: Internationalization. 
Can this be extracted to a `R.string.content_description_user_profile_avatar` ?

**Review by vitorhugods — CHANGES_REQUESTED** [review]
Oh yeah 👌🏼 🚀.

Just tiny suggestions about internationalization and resource naming. Maybe solving Detekt screams too.

File: app/src/main/kotlin/com/wire/android/ui/main/navigation/MainTopBar.kt [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+import com.wire.android.ui.common.SearchBarCollapsed
+import com.wire.android.ui.common.UserProfileAvatar
+import com.wire.android.ui.theme.WireLightColors
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+@ExperimentalAnimationApi
+@ExperimentalComposeUiApi
+@Composable
+fun MainTopBar(scope: CoroutineScope, scaffoldState: ScaffoldState, navController: NavController, hasSearchBar: Boolean = false) {
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/main/navigation/MainTopBar.kt:38]: Maybe Detekt will also complain about this method being too long.

If it was any other actual logic I would agree. But being this a UI, which could be seen as analog to XML files, this isn't bad.

Maybe yes, we could split into smaller chunks, like `@Composable fun TopBarContent()` ?

Or just disable these checks for things annotated with `Composable`. In `config/detekt/detekt.yml`, line 84:

```
    ignoreAnnotated: ['Composable']
```

File: app/src/main/kotlin/com/wire/android/ui/common/SearchBar.kt [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    matchesFound: Boolean,
+    modifier: Modifier = Modifier,
+    results: @Composable () -> Unit = {}
+) {
+    var showClearButton by remember { mutableStateOf(false) }
+    val keyboardController = LocalSoftwareKeyboardController.current
+    val focusRequester = remember { FocusRequester() }
+
+    TopAppBar(title = { Text("") },
+        backgroundColor = WireLightColors.background,
```
Comment (saleniuk) [inline app/src/main/kotlin/com/wire/android/ui/common/SearchBar.kt:70]: I think we should not refer directly to `WireLightColors`, instead use `MaterialTheme.colors.background` - we set `WireTheme` for the whole `WireApp` and this theme decides itself if it should use light or dark colors, using `MaterialTheme.colors.xxx` we get the proper one and don't have to decide again if it should be light or dark
Maybe `WireLightColors` and `WireDarkColors` should be private
I can clean this up after I complete the login screen UI, I actually thought about it, let me know if you agree 😄

**Comment by vitorhugods** [comment]
I think some lint errors were introduced by @trOnk12.

#307 was merged without passing on all checks.

File: app/src/main/kotlin/com/wire/android/ui/common/SearchBar.kt (hunk 4 above)
Comment (trOnk12) [reply app/src/main/kotlin/com/wire/android/ui/common/SearchBar.kt:70]: I agree on that :P we should try to use the MaterialTheme.color thingy where we can and if not maybe there is some way to support choosing the color depending on dark/light theme

File: app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt (hunk 2 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt:33]: Totally missed that one. Good eyes 🦅

File: app/src/main/kotlin/com/wire/android/ui/main/navigation/MainTopBar.kt (hunk 3 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/main/navigation/MainTopBar.kt:38]: I think I will add the Composable objects to this rule as usually these components tend to have long signature method headers

File: app/src/main/kotlin/com/wire/android/ui/common/SearchBar.kt (hunk 4 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/common/SearchBar.kt]: Yepp, it makes total sense. I already moved the light and dark colors palette to the Theme file and declare them as private. If you think of a better way of organising it, let me know or add it in your future PR.

**Review by vitorhugods — APPROVED** [review]
👌🏼

File: app/src/main/kotlin/com/wire/android/ui/common/SearchBar.kt [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+                    imageVector = Icons.Filled.ArrowBack,
+                    contentDescription = stringResource(id = R.string.content_description_search_back)
+                )
+            }
+        }, actions = {
+            OutlinedTextField(
+                modifier = modifier
+                    .fillMaxWidth()
+                    .onFocusChanged { focusState ->
+                        showClearButton = (focusState.isFocused)
```
Comment (borichellow) [inline app/src/main/kotlin/com/wire/android/ui/common/SearchBar.kt:83]: i think this icon should become visible not `onFocusChanged`, but when user typed something into search. So it's like "remove everything"

File: app/src/main/kotlin/com/wire/android/ui/main/navigation/MainTopBar.kt [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            modifier = Modifier.fillMaxWidth().constrainAs(topBar) {
+                top.linkTo(parent.top)
+                start.linkTo(parent.start)
+                end.linkTo(parent.end)
+            },
+            elevation = 0.dp,
+            backgroundColor = MaterialTheme.colors.background,
+            contentColor = MaterialTheme.colors.onBackground,
+            content = {
+                ConstraintLayout(Modifier.fillMaxSize()) {
```
Comment (borichellow) [inline app/src/main/kotlin/com/wire/android/ui/main/navigation/MainTopBar.kt:55]: do you really need this `ConstraintLayout` here? As far as TollBar content has a RowScope, you can just get rid of the additional ConstraintLayout and use 
`modifier = Modifier.weight(1f),
 textAlign = TextAlign.Center,` 
 for the title Text field

File: app/src/main/kotlin/com/wire/android/ui/main/userprofile/UserProfileScreen.kt [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+@Composable
+fun UserProfileScreen() {
+    UnderConstructionScreen(screenName = "UserProfileScreen")
+}
+
+@Preview(showBackground = false)
+@Composable
+fun UserProfileScreenPreview() {
+    val a: CoreLogic
```
Comment (borichellow) [inline app/src/main/kotlin/com/wire/android/ui/main/userprofile/UserProfileScreen.kt:16]: what is it?

**Review by saleniuk — COMMENTED** [review]
👍 although I have just prepared a TextField that we can also use in search so there was a small duplication of work 😅

File: app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+import androidx.constraintlayout.compose.ConstraintLayout
+import com.wire.android.R
+import com.wire.android.ui.common.Dimensions.USER_AVATAR_DEFAULT_SIZE
+
+@Composable
+fun UserProfileAvatar(avatarUrl: String, size: Dp = USER_AVATAR_DEFAULT_SIZE, modifier: Modifier = Modifier, onClick: () -> Unit) {
```
Comment (trOnk12) [inline app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt:25]: small thing but, I do not think you have to pass size:Dp = USER_AVATAR_DEFAULT_SIZE, instead you could just pass it to default modifier argument like modifier : Modifier = Modifier.padding(10.dp).size(USER_AVATAR_DEFAULT_SIZE)

**Comment by borichellow** [comment]
LGTM

**Review by borichellow — APPROVED** [review]

File: app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt (hunk 8 above)
Comment (gongracr) [reply app/src/main/kotlin/com/wire/android/ui/common/UserProfileAvatar.kt:25]: I think it is easier to allow changing the default size of the Avatar from the method signature than having to change it via the modifier parameter. But if the rest of the team agrees, we can change it later on :)

**Pull request merged**